torch/csrc/jit/tensorexpr/ir_serializer.cpp

```diff
@@ -0,0 +1,557 @@
+  if (cachedJSON(v)) {
+    return JSONOf(v);
+  }
+  v->accept(this);
+  return JSONOf(v);
+}
+
+json IRSerializer::visitAndSerialize(const std::vector<const Stmt*>& v) {
+  std::vector<json> json_vec;
+  for (const auto stmt : v) {
+    json_vec.push_back(visitAndSerialize(stmt));
+  }
+  return json_vec;
+}
+
+void torch::jit::tensorexpr::IRSerializer::visit(
+    torch::jit::tensorexpr::LongImm const* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "LongImm"},
+          {"value", v->value()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void torch::jit::tensorexpr::IRSerializer::visit(
+    torch::jit::tensorexpr::ShortImm const* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "ShortImm"},
+          {"value", v->value()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void torch::jit::tensorexpr::IRSerializer::visit(
+    torch::jit::tensorexpr::BoolImm const* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "BoolImm"},
+          {"value", v->value()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void torch::jit::tensorexpr::IRSerializer::visit(
+    torch::jit::tensorexpr::FloatImm const* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "FloatImm"},
+          {"value", v->value()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void torch::jit::tensorexpr::IRSerializer::visit(
+    torch::jit::tensorexpr::DoubleImm const* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "DoubleImm"},
+          {"value", v->value()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void torch::jit::tensorexpr::IRSerializer::visit(
+    torch::jit::tensorexpr::CharImm const* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "CharImm"},
+          {"value", v->value()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void torch::jit::tensorexpr::IRSerializer::visit(
+    torch::jit::tensorexpr::HalfImm const* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "HalfImm"},
+          {"value", static_cast<float>(v->value())},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void torch::jit::tensorexpr::IRSerializer::visit(
+    torch::jit::tensorexpr::ByteImm const* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "ByteImm"},
+          {"value", v->value()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void torch::jit::tensorexpr::IRSerializer::visit(
+    torch::jit::tensorexpr::IntImm const* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "IntImm"},
+          {"value", v->value()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
```

Comment:
Any reason not to use the DTYPE macro for these identical looking definitions?

Reply:
You need the manual cast from `at::Half`, if theres a good way of using macro for all except that lmk

Reply:
There is!
https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/ir.h#L334 has an example of a macro that hits all scalar types + Half, but not Bool. It would be easy to switch this up to include Bool but not Half.

Reply:
I tried to do this, and ran into some weird macro-related issues I had a hard time figuring it out. I can try again if you think it's important.